AfterClass-io/afterclass.io-v2
# chore(tests): introduce cypress e2e testing

This change introduces some basic e2e tests for auth and reviews modules. Specifics can be found in the cypress/e2e folders

## Changes

1. add github workflow using cypress github action with `bun`
2. add `data-test=*` attributes to various components
3. revert posthog middleware -- previously broke review submissions, stuck in infinite login prompt
4. add symlink from `prisma/data` -> `cypress/fixtures/prisma` such that mock database data is available as fixtures for cypress
5. update db seed command to use `bun`. `ts-node` no longer works well, see `https://github.com/TypeStrong/ts-node/issues/2100`


## E2E Test Summary

```
  1-auth/login.cy.js
  Login
    Successful Login
      ✓ should login a user with valid v1 credentials (2375ms)
      ✓ should login a user with valid v2 credentials (2151ms)
      ✓ should be able to navigate to login page and login (1982ms)
    Unsuccessful Logins
      ✓ should warn user to fill in email (433ms)
      ✓ should warn user to fill in password (599ms)
      ✓ should warn user to fill in email and password (213ms)
      ✓ should not login a user with invalid credentials (1932ms)

  1-auth/register.cy.js
  Register
    Successful Registration
      ✓ should register a user with credentials (2451ms)
      ✓ should be able to navigate to register page and register (2559ms)
    Incomplete Registrations
      ✓ should warn user to fill in fields (233ms)
      ✓ should warn user to fill in schoool email (904ms)

  1-auth/reset_pwd.cy.js
  ResetPwd
    Successful Password Reset
      ✓ should be able to submit password reset request (1894ms)
      ✓ should be able to navigate to register page and submit password reset request (2709ms)
    Incomplete Password Reset
      ✓ should warn user to fill in fields (226ms)
      ✓ should warn user to fill in schoool email (552ms)

  2-reviews/course.cy.js
  Home
    Basic Navigations
      ✓ should be able to navigate to login page (902ms)
      ✓ should be able to navigate to bid analytics page (486ms)
      ✓ should be able to navigate to reviews page (856ms)
      ✓ should be able to navigate to professor reviews page (591ms)
    Unauthenticated User
      ✓ should be able to see login overlays (592ms)
      ✓ should not be able to see reviews ratings (297ms)
      ✓ should not be able to open information modal (355ms)
      ✓ should not be able to filter reviews (330ms)
      ✓ should not be able to open review modal (554ms)
      ✓ should not be able to load more reviews (2407ms)
      ✓ should not be able to like a review (1380ms)
      ✓ should not be able to navigate to review submission (401ms)
    Authenticated User
      ✓ should not be able to see login overlays (2383ms)
      ✓ should be able to see reviews ratings (1902ms)
      ✓ should be able to open information modal (2473ms)
      ✓ should be able to filter reviews (2523ms)
      ✓ should be able to open review modal (2525ms)
      ✓ should be able to like a review (2635ms)
      ✓ should be able to unlike a review (2542ms)
      ✓ should be able to load more reviews (3390ms)
      ✓ should be able to write a review (2566ms)
    Data Accuracy
      ✓ should display accurate course information (2870ms)
      ✓ should display accurate professor information (10720ms)
      ✓ should display accurate review ratings (2455ms)
      ✓ should display accurate review counts (2594ms)

  2-reviews/home.cy.js
  Home
    Basic Navigations
      ✓ should be able to navigate to login page and login (704ms)
      ✓ should be able to navigate to bid analytics page (326ms)
      ✓ should be able to navigate to course reviews page (702ms)
      ✓ should be able to navigate to professor reviews page (684ms)
    Unauthenticated User
      ✓ should be able to see login overlay on review item (464ms)
      ✓ should not be able to open review modal (558ms)
      ✓ should not be able to load more reviews (2285ms)
      ✓ should not be able to like a review (1414ms)
      ✓ should not be able to navigate to review submission (444ms)
    Authenticated User
      ✓ should not be able to see login overlay on review item (2375ms)
      ✓ should be able to open review modal (2599ms)
      ✓ should be able to like a review (2557ms)
      ✓ should be able to unlike a review (1986ms)
      ✓ should be able to load more reviews (2903ms)
      ✓ should be able to write a review (2522ms)

  2-reviews/professor.cy.js
  Home
    Basic Navigations
      ✓ should be able to navigate to login page (737ms)
      ✓ should be able to navigate to bid analytics page (427ms)
      ✓ should be able to navigate to reviews page (772ms)
      ✓ should be able to navigate to course reviews page (639ms)
    Unauthenticated User
      ✓ should be able to see login overlays (436ms)
      ✓ should not be able to see reviews ratings (157ms)
      ✓ should not be able to filter reviews (160ms)
      ✓ should not be able to open review modal (505ms)
      ✓ should not be able to load more reviews (2295ms)
      ✓ should not be able to like a review (1340ms)
      ✓ should not be able to navigate to review submission (362ms)
    Authenticated User
      ✓ should not be able to see login overlays (2433ms)
      ✓ should be able to see reviews ratings (2281ms)
      ✓ should be able to filter reviews (2473ms)
      ✓ should be able to open review modal (2572ms)
      ✓ should be able to like a review (2566ms)
      ✓ should be able to unlike a review (2534ms)
      ✓ should be able to load more reviews (2901ms)
      ✓ should be able to write a review (2502ms)
    Data Accuracy
      ✓ should display accurate professor information (1888ms)
      ✓ should display accurate course information (4504ms)
      ✓ should display accurate review ratings (1913ms)
      ✓ should display accurate review counts (2565ms)

  2-reviews/search.cy.js
  Home
    Search Functionality on Home Page
      ✓ should bring up searchbar when shortcut or search input is clicked (1935ms)
      ✓ should be able to search for a course (1776ms)
      ✓ should be able to search for a professor (1743ms)
      ✓ should not be able to sql inject (2299ms)
      ✓ should display not-found if no search results (1834ms)
    Search Functionality on Search Page
      ✓ should bring up searchbar when shortcut or search input is clicked (1485ms)
      ✓ should be able to search for a course (1505ms)
      ✓ should be able to search for a professor (1557ms)
      ✓ should not be able to sql inject (2053ms)
      ✓ should display not-found if no search results (1675ms)
    Search Functionality on Course Page
      ✓ should bring up searchbar when shortcut or search input is clicked (1646ms)
      ✓ should be able to search for a course (1578ms)
      ✓ should be able to search for a professor (1577ms)
      ✓ should not be able to sql inject (1926ms)
      ✓ should display not-found if no search results (1829ms)
    Search Functionality on Professor Page
      ✓ should bring up searchbar when shortcut or search input is clicked (1571ms)
      ✓ should be able to search for a course (1602ms)
      ✓ should be able to search for a professor (1565ms)
      ✓ should not be able to sql inject (2003ms)
      ✓ should display not-found if no search results (1757ms)
  
  2-reviews/submit.cy.js
  Home
    Basic Navigations
      ✓ should be able to navigate to bid analytics page (1504ms)
      ✓ should be able to navigate to reviews page (1197ms)
    Review Submission
      ✓ should be able to submit a course review (4673ms)
      ✓ should be able to submit a course review anonymously (4279ms)
      ✓ should be able to submit a course & professor review (7263ms)
      ✓ should be able to submit a course & professor review anonymously (7793ms)  
```
